theano/gpuarray/tests/test_linalg.py

```diff
@@ -293,3 +302,132 @@ def test_gpu_singular_values(self):
+        return fn(A_val)
+
+    def check_gpu_qr(self, M, N, complete=True, rtol=None, atol=None):
+        A = rand(M, N).astype('float32')
+        if complete:
+            Q_gpu, R_gpu = self.run_gpu_qr(A, complete=complete)
+        else:
+            R_gpu = self.run_gpu_qr(A, complete=complete)
+
+        Q_np, R_np = np.linalg.qr(A, mode='reduced')
+        utt.assert_allclose(R_gpu, R_np, rtol=rtol, atol=atol)
+        if complete:
+            utt.assert_allclose(Q_gpu, Q_np, rtol=rtol, atol=atol)
+
+    def test_gpu_qr(self):
+        self.check_gpu_qr(1000, 500, atol=1e-3)
+        self.check_gpu_qr(1000, 500, complete=False, atol=1e-3)
+        self.check_gpu_qr(500, 1000, atol=1e-3)
+        self.check_gpu_qr(500, 1000, complete=False, atol=1e-3)
+
+    def test_gpu_qr_opt(self):
+        A = theano.tensor.fmatrix("A")
+        fn = theano.function([A], qr(A), mode=mode_with_gpu)
+        assert any([
+            isinstance(node.op, GpuMagmaQR) and node.op.complete
+            for node in fn.maker.fgraph.toposort()
+        ])
+
+    def test_gpu_qr_incomplete_opt(self):
+        A = theano.tensor.fmatrix("A")
+        fn = theano.function([A], qr(A, mode='r'), mode=mode_with_gpu)
+        assert any([
+            isinstance(node.op, GpuMagmaQR) and not node.op.complete
+            for node in fn.maker.fgraph.toposort()
+        ])
+
+    def run_gpu_eigh(self, A_val, UPLO='L', compute_v=True):
+        A = theano.tensor.fmatrix("A")
+        fn = theano.function([A], GpuMagmaEigh(UPLO=UPLO, compute_v=compute_v)(A),
+                             mode=mode_with_gpu)
+        return fn(A_val)
+
+    def check_gpu_eigh(self, N, UPLO='L', compute_v=True, rtol=None, atol=None):
+        A = rand(N, N).astype('float32')
+        A = np.dot(A.T, A)
+        d_np, v_np = np.linalg.eigh(A, UPLO=UPLO)
+        if compute_v:
+            d_gpu, v_gpu = self.run_gpu_eigh(A, UPLO=UPLO, compute_v=compute_v)
+        else:
+            d_gpu = self.run_gpu_eigh(A, UPLO=UPLO, compute_v=False)
+        utt.assert_allclose(d_gpu, d_np, rtol=rtol, atol=atol)
+        if compute_v:
+            utt.assert_allclose(
+                np.dot(v_gpu, v_gpu.T), np.eye(N), rtol=rtol, atol=atol)
+            D_m = np.zeros_like(A)
+            np.fill_diagonal(D_m, d_gpu)
+            utt.assert_allclose(
+                np.dot(np.dot(v_gpu, D_m), v_gpu.T), A, rtol=rtol, atol=atol)
+
+    def test_gpu_eigh(self):
+        self.check_gpu_eigh(1000, UPLO='L', atol=1e-3)
+        self.check_gpu_eigh(1000, UPLO='U', atol=1e-3)
+        self.check_gpu_eigh(1000, UPLO='L', compute_v=False, atol=1e-3)
+        self.check_gpu_eigh(1000, UPLO='U', compute_v=False, atol=1e-3)
+
+    def test_gpu_eigh_opt(self):
+        A = theano.tensor.fmatrix("A")
+        fn = theano.function([A], eigh(A), mode=mode_with_gpu)
+        assert any([
+            isinstance(node.op, GpuMagmaEigh)
+            for node in fn.maker.fgraph.toposort()
+        ])
```

Comment:
For opt tests (e.g. `test_gpu_cholesky_opt`), add also tests for float16 (e.g. `test_gpu_cholesky_opt_f16`).